README.md
# Layoutry

<p align="center">
 <br>
 <img width="400" src="https://user-images.githubusercontent.com/59066341/211650854-e7350bd8-aba9-430b-a216-057c126e7354.svg" alt="Package Logo">
 <br>
 <br>
 <a href="https://pub.dev/packages/layoutry">
  <img src="https://img.shields.io/pub/v/layoutry?color=blue" alt="pub version" />
 </a>
  <a href="https://github.com/insolite-dev/layoutry/blob/main/LICENSE">
    <img src="https://img.shields.io/badge/License-MIT-red.svg"
      alt="License: MIT" />
  </a>
 <br>
  <a href="https://discord.gg/CtStkzrHV3">
    <img src="https://img.shields.io/discord/914899238415130714?color=blue&label=Insolite Community&logo=discord"
      alt="Insolite Community Discord Invite Link" />
  </a>   
</p>

## Installing
**See the official installing guidline from [layoutry/install](https://pub.dev/packages/layoutry/install)**

## Usage & Overview

<img width="1000" src="https://user-images.githubusercontent.com/59066341/211651068-27f5ca62-4d0d-44fa-b87d-7924605c017c.gif" alt="Package Example Overview">

Layoutry is a modified implementation of the **LayoutBuilder**, so that we can use its **LayoutInfo** to easily catch the device's type via its screen size.
Where the **LayoutDevice** is a custom `Platform.operatingSystem` implementation `enum`, which used to mark current device's type by screen size.

Here is the code:
```dart
Layoutry(
  builder: (context, info) {
    // A manual defined layout-device to color map.
    final colors = <LayoutDevice, Color>{
      LayoutDevice.mobile: Colors.blue,
      LayoutDevice.tablet: Colors.red,
      LayoutDevice.web: Colors.green,
    };

    return AnimatedContainer(
      duration: const Duration(milliseconds: 500),
      // color will be generated from [colors],
      // by listening to device's screen size.
      color: colors[info.device],
      child: Center(child: Builder(builder: (context) {
        // If the device's screen size is like mobile:
        // "Hi Mobile" will written in the screen.
        if (info.device.isMobile()) {
          return const Text(
            'Hi Mobile',
            style: TextStyle(color: Colors.white, fontSize: 20),
          );
        }

        // If the device's screen size is like web:
        // "Hi Web" will written in the screen.
        if (info.device.isWeb()) {
          return const Text(
            'Hi Web',
            style: TextStyle(color: Colors.white, fontSize: 20),
          );
        }

        // In other cases: Hi current device's screen size
        // appropriate device type will be written in the screen.
        return Text(
          'Hi ${info.device.toString()}',
          style: const TextStyle(color: Colors.white, fontSize: 20),
        );
      })),
    );
  },
)
```


CHANGELOG.md
## 1.0.0

# <img width="100" src="https://user-images.githubusercontent.com/59066341/211650854-e7350bd8-aba9-430b-a216-057c126e7354.svg" alt="Package Logo"> v1.0.0

 **The very first version of layoutry.**

<img width="500" src="https://user-images.githubusercontent.com/59066341/211651068-27f5ca62-4d0d-44fa-b87d-7924605c017c.gif" alt="Package Example Overview">


test/layout_info_test.dart
//
// Copyright 2023 present Insolite. All rights reserved.
// Use of this source code is governed by MIT license
// that can be found in the LICENSE file.
//

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:layoutry/layout_info.dart';

void main() {
  group('LayoutDeviceExtension:', () {
    test('isMobile', () {
      final tests = <LayoutDevice, bool>{
        LayoutDevice.mobile: true,
        LayoutDevice.tablet: false,
        LayoutDevice.web: false,
        LayoutDevice.webHD: false,
      };

      tests.forEach((t, expected) => expect(t.isMobile(), expected));
    });

    test('isTablet', () {
      final tests = <LayoutDevice, bool>{
        LayoutDevice.mobile: false,
        LayoutDevice.tablet: true,
        LayoutDevice.web: false,
        LayoutDevice.webHD: false,
      };

      tests.forEach((t, expected) => expect(t.isTablet(), expected));
    });

    test('isWeb', () {
      final tests = <LayoutDevice, bool>{
        LayoutDevice.mobile: false,
        LayoutDevice.tablet: false,
        LayoutDevice.web: true,
        LayoutDevice.webHD: false,
      };

      tests.forEach((t, expected) => expect(t.isWeb(), expected));
    });

    test('isWebHD', () {
      final tests = <LayoutDevice, bool>{
        LayoutDevice.mobile: false,
        LayoutDevice.tablet: false,
        LayoutDevice.web: false,
        LayoutDevice.webHD: true,
      };

      tests.forEach((t, expected) => expect(t.isWebHD(), expected));
    });
  });

  group('LayoutInfo:', () {
    test('buildFromConstraints', () {
      final tests = <BoxConstraints, LayoutInfo>{
        const BoxConstraints(maxWidth: 200): const LayoutInfo(
          device: LayoutDevice.mobile,
          constraints: BoxConstraints(maxWidth: 200),
        ),
        const BoxConstraints(maxWidth: 1200): const LayoutInfo(
          device: LayoutDevice.tablet,
          constraints: BoxConstraints(maxWidth: 1200),
        ),
        const BoxConstraints(maxWidth: 1925): const LayoutInfo(
          device: LayoutDevice.web,
          constraints: BoxConstraints(maxWidth: 1925),
        ),
        const BoxConstraints(maxWidth: 100000): const LayoutInfo(
          device: LayoutDevice.webHD,
          constraints: BoxConstraints(maxWidth: 100000),
        ),
      };

      tests.forEach((t, expected) {
        final got = LayoutInfo.buildFromConstraints(t);

        expect(got.device, expected.device);
        expect(got.constraints, expected.constraints);
      });
    });

    test('deviceFromWeight', () {
      final tests = <int, LayoutDevice>{
        800: LayoutDevice.mobile,
        1200: LayoutDevice.tablet,
        1600: LayoutDevice.web,
        1925: LayoutDevice.web,
        100000: LayoutDevice.webHD,
      };

      tests.forEach(
        (t, expected) => expect(
          LayoutInfo.deviceFromWeight(t.toDouble()),
          expected,
        ),
      );
    });
  });
}


lib/layoutry.dart
//
// Copyright 2023 present Insolite. All rights reserved.
// Use of this source code is governed by MIT license
// that can be found in the LICENSE file.
//

library layoutry;

import 'package:flutter/material.dart';
import 'package:layoutry/layout_info.dart';

export 'layout_info.dart';

/// A builder type definition for the [Layoutry].
/// Requires a [BuildContext] and [LayoutInfo].
typedef LayoutryBuilder = Widget Function(
  BuildContext context,
  LayoutInfo info,
);

/// Layoutry is a modified [LayoutBuilder], which comes with a parsed [LayoutInfo].
/// Where you can use this info to check the device type of current running screen.
///
/// For example:
/// ```dart
///  Layoutry(builder: (context, info) {
///    if(info.device.isMobile()) {
///      return const Text('Hi Mobile');
///    }
///
///    if(info.device.isWeb()) {
///      return const Text('Hi Web');
///    }
///
///    return Text('Hi ${info.device.toString()}');
///  });
/// ```
class Layoutry extends StatelessWidget {
  final LayoutryBuilder builder;
  const Layoutry({Key? key, required this.builder}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) => builder(
        context,
        LayoutInfo.buildFromConstraints(constraints),
      ),
    );
  }
}


lib/layout_info.dart
//
// Copyright 2023 present Insolite. All rights reserved.
// Use of this source code is governed by MIT license
// that can be found in the LICENSE file.
//

import 'package:flutter/material.dart';

// TODO: Has to be re-calculated ~ exact ~ widths.
const int widthMobile = 800;
const int widthTablet = 1200;
const int widthFloatingWindow = 1600;
const int widthFullHD = 1925;
const int widthFullHDPlus = 100000;

/// A device type by its screen size.
///
/// Similar to `Platform.operatingSystem` from `dart:io`.
enum LayoutDevice { mobile, tablet, web, webHD } // TODO: should remove webHD.

extension LayoutDeviceExtension on LayoutDevice {
  bool isMobile() => this == LayoutDevice.mobile;
  bool isTablet() => this == LayoutDevice.tablet;
  bool isWeb() => this == LayoutDevice.web;
  bool isWebHD() => this == LayoutDevice.webHD;
}

/// A main class of information of [Layoutry].
/// Includes the default [BoxConstraints] (of the [LayoutBuilder]), and [LayoutDevice].
class LayoutInfo {
  final BoxConstraints constraints;
  final LayoutDevice device;

  const LayoutInfo({required this.constraints, required this.device});

  /// Generates [LayoutInfo] from the [BoxConstraints] of the [LayoutBuilder].
  /// Uses [deviceByWeigth] to catch the right [LayoutDevice] by `maxWidth`.
  static LayoutInfo buildFromConstraints(BoxConstraints c) {
    final width = c.maxWidth;
    return LayoutInfo(constraints: c, device: deviceFromWeight(width));
  }

  /// Generates a [LayoutDevice] by [maxWidth] of [BoxConstraints].
  // TODO: withFloatingWindow shouldn't return [LayoutDevice.web].
  static LayoutDevice deviceFromWeight(double w) {
    if (w <= widthMobile) return LayoutDevice.mobile;
    if (w <= widthTablet) return LayoutDevice.tablet;
    if (w <= widthFullHD || w <= widthFloatingWindow) {
      return LayoutDevice.web;
    }

    if (w <= widthFullHDPlus) return LayoutDevice.webHD;

    // Could be any.
    return LayoutDevice.mobile;
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:layoutry/layoutry.dart';

void main() => runApp(const App());

class App extends StatelessWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) => const MaterialApp(home: Home());
}

// See overview at:
// https://user-images.githubusercontent.com/59066341/211651068-27f5ca62-4d0d-44fa-b87d-7924605c017c.gif
class Home extends StatelessWidget {
  const Home({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Layoutry(
        builder: (context, info) {
          // A manual defined layout-device to color map.
          final colors = <LayoutDevice, Color>{
            LayoutDevice.mobile: Colors.blue,
            LayoutDevice.tablet: Colors.red,
            LayoutDevice.web: Colors.green,
          };

          return AnimatedContainer(
            duration: const Duration(milliseconds: 500),
            // color will be generated from [colors],
            // by listening to device's screen size.
            color: colors[info.device],
            child: Center(child: Builder(builder: (context) {
              // If the device's screen size is like mobile:
              // "Hi Mobile" will written in the screen.
              if (info.device.isMobile()) {
                return const Text(
                  'Hi Mobile',
                  style: TextStyle(color: Colors.white, fontSize: 20),
                );
              }

              // If the device's screen size is like web:
              // "Hi Web" will written in the screen.
              if (info.device.isWeb()) {
                return const Text(
                  'Hi Web',
                  style: TextStyle(color: Colors.white, fontSize: 20),
                );
              }

              // In other cases: Hi current device's screen size
              // appropriate device type will be written in the screen.
              return Text(
                'Hi ${info.device.toString()}',
                style: const TextStyle(color: Colors.white, fontSize: 20),
              );
            })),
          );
        },
      ),
    );
  }
}
